README.md
## utools-chatroom

在线聊天室

[live demo](http://39.105.91.194:8090/)

TODO
- [x] 调整默认大小
- [x] 去掉底部栏
- [ ] 重写 drop 和 drag  事件
- [ ] 重写 parse 事件
- [x] 自动重试,手动重试
- [x] 重试一次
- [x] 消息提醒
- [ ] 图片粘贴上传


utools-chatroom/CHANGELOG.md
### 更新日志

### 0.0.9

- 新增消息提醒

#### 0.0.8

- 更换聊天室图标

#### 0.0.7

- 增加自动重连和手动重连功能

#### 0.0.6

- 网上畅聊，可以发送文字，表情，或者图片。


static/js/emotion.js
(function () {
	var QxEmotion = function (_emotion_btn, _edit) {
        QxEmotion.Bind(_emotion_btn, _edit);
	};

    QxEmotion.Bind = function (_emotion_btn, _edit) {
        _emotion_btn.on('click', function(){
            var emotion_panel = $('.emotion-panel');
            if(0 == emotion_panel.length) {
                emotion_panel = QxEmotion._CreateEmotionPanel();
                //挂接click事件
                emotion_panel.find('td').on('click', function () {
                    emotion_panel.hide(60);
                    var emotion_name = $(this).children('img').attr('alt');
                    _edit.val(_edit.val() + '[#' + emotion_name + ']');
                    _edit.focus();
                });
                return setTimeout(function () {QxEmotion._ToggleEmotionPanel(emotion_panel, _emotion_btn);}, 100);
            } else {
                QxEmotion._ToggleEmotionPanel(emotion_panel, _emotion_btn);
            }
        });
    };

	QxEmotion.Parse = function (_text) {
		var re_ret;
		var re = new RegExp("[[]#(.+?)]","gm"); //js的正则真TMD操蛋..
		var emotion_name = '';
		var emotion_path = '';
		while(null != (re_ret = re.exec(_text))){
			emotion_name = re_ret[1];
			if(undefined != QxEmotion.data[emotion_name]){
				emotion_path = QxEmotion.path + QxEmotion.data[emotion_name];
				_text = _text.replace(re_ret[0], '<img src="' + emotion_path + '" atl="' + emotion_name + '">');
			}
		}
		return _text;
	};

	QxEmotion.path = './static/images/qqface/';

	QxEmotion.data = {
		'微笑' : '36.gif',
		'亲亲' : '3.gif',
		'流泪' : '41.gif',
		'发呆' : '78.gif',
		'调皮' : '44.gif',
		'阴险' : '40.gif',
		'大笑' : '85.gif',
		'晕倒' : '96.gif',
		'白眼' : '80.gif',
		'脸红' : '75.gif',
		'妩媚' : '15.gif',
		'疑问' : '1.gif',
		'尴尬' : '4.gif',
		'抠鼻' : '99.gif',
		'流汗' : '98.gif',
		'再见' : '97.gif',
		'困'   : '8.gif',
		'衰'   : '53.gif',
		'抓狂' : '9.gif',
		'奋斗' : '10.gif',
		'ok'   : '11.gif',
		'折磨' : '12.gif',
		'委屈' : '13.gif',
		'嘘'   : '14.gif',
		'强'   : '43.gif',
		'饥饿' : '17.gif',
		'闭嘴' : '18.gif',
		'爱你' : '19.gif',
		'哭泣' : '70.gif',
		'口罩' : '20.gif',
		'猪头' : '21.gif',
		'难过' : '22.gif',
		'鄙视' : '23.gif',
		'抱抱' : '45.gif',
		'鼓掌' : '69.gif',
	};

	QxEmotion._CreateEmotionPanel = function () {
		//创建panel
		var emotion_html = '<div class="panel panel-default emotion-panel">' +
			'<div class="panel-body"><table class="table-condensed table-bordered"><tbody>';
		var column_count = 0;
		for(var emotion_name in QxEmotion.data){
			//创建tr
			if(column_count == 0) {
				emotion_html += '<tr>';
			}
			//生成img
			var path = QxEmotion.path + QxEmotion.data[emotion_name];
			emotion_html += '<td><img src="'+ path + '" alt="' + emotion_name + '"></td>';
			column_count++;
			//闭合tr
			if(column_count == 7){ //这里可以设定列数
				emotion_html += '</tr>';
				column_count = 0;
			}
		}
		//闭合panel
		emotion_html += '</tbody></table></div></div>';
		$('body').append(emotion_html);

		return $('.emotion-panel');
	};

    QxEmotion._ToggleEmotionPanel = function (_emotion_panel, _emotion_btn) {
        if (_emotion_panel.is(':visible')) {
            _emotion_panel.hide(60);
        } else {
            _emotion_panel.css("left", _emotion_btn.offset().left);
            _emotion_panel.css("top", _emotion_btn.offset().top - _emotion_panel.height() - 5);
            _emotion_panel.show(60);
        }
    };

	window.QxEmotion = QxEmotion;
})();

utools-chatroom/static/js/reconnecting-websocket.js
// MIT License:
//
// Copyright (c) 2010-2012, Joe Walnes
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

/**
 * This behaves like a WebSocket in every way, except if it fails to connect,
 * or it gets disconnected, it will repeatedly poll until it successfully connects
 * again.
 *
 * It is API compatible, so when you have:
 *   ws = new WebSocket('ws://....');
 * you can replace with:
 *   ws = new ReconnectingWebSocket('ws://....');
 *
 * The event stream will typically look like:
 *  onconnecting
 *  onopen
 *  onmessage
 *  onmessage
 *  onclose // lost connection
 *  onconnecting
 *  onopen  // sometime later...
 *  onmessage
 *  onmessage
 *  etc...
 *
 * It is API compatible with the standard WebSocket API, apart from the following members:
 *
 * - `bufferedAmount`
 * - `extensions`
 * - `binaryType`
 *
 * Latest version: https://github.com/joewalnes/reconnecting-websocket/
 * - Joe Walnes
 *
 * Syntax
 * ======
 * var socket = new ReconnectingWebSocket(url, protocols, options);
 *
 * Parameters
 * ==========
 * url - The url you are connecting to.
 * protocols - Optional string or array of protocols.
 * options - See below
 *
 * Options
 * =======
 * Options can either be passed upon instantiation or set after instantiation:
 *
 * var socket = new ReconnectingWebSocket(url, null, { debug: true, reconnectInterval: 4000 });
 *
 * or
 *
 * var socket = new ReconnectingWebSocket(url);
 * socket.debug = true;
 * socket.reconnectInterval = 4000;
 *
 * debug
 * - Whether this instance should log debug messages. Accepts true or false. Default: false.
 *
 * automaticOpen
 * - Whether or not the websocket should attempt to connect immediately upon instantiation. The socket can be manually opened or closed at any time using ws.open() and ws.close().
 *
 * reconnectInterval
 * - The number of milliseconds to delay before attempting to reconnect. Accepts integer. Default: 1000.
 *
 * maxReconnectInterval
 * - The maximum number of milliseconds to delay a reconnection attempt. Accepts integer. Default: 30000.
 *
 * reconnectDecay
 * - The rate of increase of the reconnect delay. Allows reconnect attempts to back off when problems persist. Accepts integer or float. Default: 1.5.
 *
 * timeoutInterval
 * - The maximum time in milliseconds to wait for a connection to succeed before closing and retrying. Accepts integer. Default: 2000.
 *
 */
(function (global, factory) {
    if (typeof define === 'function' && define.amd) {
        define([], factory);
    } else if (typeof module !== 'undefined' && module.exports){
        module.exports = factory();
    } else {
        global.ReconnectingWebSocket = factory();
    }
})(this, function () {

    if (!('WebSocket' in window)) {
        return;
    }

    function ReconnectingWebSocket(url, protocols, options) {

        // Default settings
        var settings = {

            /** Whether this instance should log debug messages. */
            debug: false,

            /** Whether or not the websocket should attempt to connect immediately upon instantiation. */
            automaticOpen: true,

            /** The number of milliseconds to delay before attempting to reconnect. */
            reconnectInterval: 1000,
            /** The maximum number of milliseconds to delay a reconnection attempt. */
            maxReconnectInterval: 30000,
            /** The rate of increase of the reconnect delay. Allows reconnect attempts to back off when problems persist. */
            reconnectDecay: 1.5,

            /** The maximum time in milliseconds to wait for a connection to succeed before closing and retrying. */
            timeoutInterval: 2000,

            /** The maximum number of reconnection attempts to make. Unlimited if null. */
            maxReconnectAttempts: null,

            /** The binary type, possible values 'blob' or 'arraybuffer', default 'blob'. */
            binaryType: 'blob'
        }
        if (!options) { options = {}; }

        // Overwrite and define settings with options if they exist.
        for (var key in settings) {
            if (typeof options[key] !== 'undefined') {
                this[key] = options[key];
            } else {
                this[key] = settings[key];
            }
        }

        // These should be treated as read-only properties

        /** The URL as resolved by the constructor. This is always an absolute URL. Read only. */
        this.url = url;

        /** The number of attempted reconnects since starting, or the last successful connection. Read only. */
        this.reconnectAttempts = 0;

        /**
         * The current state of the connection.
         * Can be one of: WebSocket.CONNECTING, WebSocket.OPEN, WebSocket.CLOSING, WebSocket.CLOSED
         * Read only.
         */
        this.readyState = WebSocket.CONNECTING;

        /**
         * A string indicating the name of the sub-protocol the server selected; this will be one of
         * the strings specified in the protocols parameter when creating the WebSocket object.
         * Read only.
         */
        this.protocol = null;

        // Private state variables

        var self = this;
        var ws;
        var forcedClose = false;
        var timedOut = false;
        var eventTarget = document.createElement('div');

        // Wire up "on*" properties as event handlers

        eventTarget.addEventListener('open',       function(event) { self.onopen(event); });
        eventTarget.addEventListener('close',      function(event) { self.onclose(event); });
        eventTarget.addEventListener('connecting', function(event) { self.onconnecting(event); });
        eventTarget.addEventListener('totalclose', function (event) { self.ontotalclose(event); })
        eventTarget.addEventListener('message',    function(event) { self.onmessage(event); });
        eventTarget.addEventListener('error',      function(event) { self.onerror(event); });

        // Expose the API required by EventTarget

        this.addEventListener = eventTarget.addEventListener.bind(eventTarget);
        this.removeEventListener = eventTarget.removeEventListener.bind(eventTarget);
        this.dispatchEvent = eventTarget.dispatchEvent.bind(eventTarget);

        /**
         * This function generates an event that is compatible with standard
         * compliant browsers and IE9 - IE11
         *
         * This will prevent the error:
         * Object doesn't support this action
         *
         * http://stackoverflow.com/questions/19345392/why-arent-my-parameters-getting-passed-through-to-a-dispatched-event/19345563#19345563
         * @param s String The name that the event should use
         * @param args Object an optional object that the event will use
         */
        function generateEvent(s, args) {
        	var evt = document.createEvent("CustomEvent");
        	evt.initCustomEvent(s, false, false, args);
        	return evt;
        };

        this.open = function (reconnectAttempt) {
            ws = new WebSocket(self.url, protocols || []);
            ws.binaryType = this.binaryType;

            if (reconnectAttempt) {
                if (this.maxReconnectAttempts && this.reconnectAttempts > this.maxReconnectAttempts) {
                    eventTarget.dispatchEvent(generateEvent('totalclose'));
                    return;
                }
            } else {
                eventTarget.dispatchEvent(generateEvent('connecting'));
                this.reconnectAttempts = 0;
            }

            if (self.debug || ReconnectingWebSocket.debugAll) {
                console.log('ReconnectingWebSocket', 'attempt-connect', self.url);
            }

            var localWs = ws;
            var timeout = setTimeout(function() {
                if (self.debug || ReconnectingWebSocket.debugAll) {
                    console.log('ReconnectingWebSocket', 'connection-timeout', self.url);
                }
                timedOut = true;
                localWs.close();
                timedOut = false;
            }, self.timeoutInterval);

            ws.onopen = function(event) {
                clearTimeout(timeout);
                if (self.debug || ReconnectingWebSocket.debugAll) {
                    console.log('ReconnectingWebSocket', 'onopen', self.url);
                }
                self.protocol = ws.protocol;
                self.readyState = WebSocket.OPEN;
                self.reconnectAttempts = 0;
                var e = generateEvent('open');
                e.isReconnect = reconnectAttempt;
                reconnectAttempt = false;
                eventTarget.dispatchEvent(e);
            };

            ws.onclose = function(event) {
                clearTimeout(timeout);
                ws = null;
                if (forcedClose) {
                    self.readyState = WebSocket.CLOSED;
                    eventTarget.dispatchEvent(generateEvent('close'));
                } else {
                    self.readyState = WebSocket.CONNECTING;
                    var e = generateEvent('connecting');
                    e.code = event.code;
                    e.reason = event.reason;
                    e.wasClean = event.wasClean;
                    eventTarget.dispatchEvent(e);
                    if (!reconnectAttempt && !timedOut) {
                        if (self.debug || ReconnectingWebSocket.debugAll) {
                            console.log('ReconnectingWebSocket', 'onclose', self.url);
                        }
                        eventTarget.dispatchEvent(generateEvent('close'));
                    }

                    var timeout = self.reconnectInterval * Math.pow(self.reconnectDecay, self.reconnectAttempts);
                    setTimeout(function() {
                        self.reconnectAttempts++;
                        self.open(true);
                    }, timeout > self.maxReconnectInterval ? self.maxReconnectInterval : timeout);
                }
            };
            ws.onmessage = function(event) {
                if (self.debug || ReconnectingWebSocket.debugAll) {
                    console.log('ReconnectingWebSocket', 'onmessage', self.url, event.data);
                }
                var e = generateEvent('message');
                e.data = event.data;
                eventTarget.dispatchEvent(e);
            };
            ws.onerror = function(event) {
                if (self.debug || ReconnectingWebSocket.debugAll) {
                    console.log('ReconnectingWebSocket', 'onerror', self.url, event);
                }
                eventTarget.dispatchEvent(generateEvent('error'));
            };
        }

        // Whether or not to create a websocket upon instantiation
        if (this.automaticOpen == true) {
            this.open(false);
        }

        /**
         * Transmits data to the server over the WebSocket connection.
         *
         * @param data a text string, ArrayBuffer or Blob to send to the server.
         */
        this.send = function(data) {
            if (ws) {
                if (self.debug || ReconnectingWebSocket.debugAll) {
                    console.log('ReconnectingWebSocket', 'send', self.url, data);
                }
                return ws.send(data);
            } else {
                throw 'INVALID_STATE_ERR : Pausing to reconnect websocket';
            }
        };

        /**
         * Closes the WebSocket connection or connection attempt, if any.
         * If the connection is already CLOSED, this method does nothing.
         */
        this.close = function(code, reason) {
            // Default CLOSE_NORMAL code
            if (typeof code == 'undefined') {
                code = 1000;
            }
            forcedClose = true;
            if (ws) {
                ws.close(code, reason);
            }
        };

        /**
         * Additional public API method to refresh the connection if still open (close, re-open).
         * For example, if the app suspects bad data / missed heart beats, it can try to refresh.
         */
        this.refresh = function() {
            if (ws) {
                ws.close();
            }
        };
    }

    /**
     * An event listener to be called when the WebSocket connection's readyState changes to OPEN;
     * this indicates that the connection is ready to send and receive data.
     */
    ReconnectingWebSocket.prototype.onopen = function(event) {};
    /** An event listener to be called when the WebSocket connection's readyState changes to CLOSED. */
    ReconnectingWebSocket.prototype.onclose = function(event) {};
    /** An event listener to be called when a connection begins being attempted. */
    ReconnectingWebSocket.prototype.onconnecting = function(event) {};
    /** An event listener to be called when a message is received from the server. */
    ReconnectingWebSocket.prototype.onmessage = function(event) {};
    /** An event listener to be called when an error occurs. */
    ReconnectingWebSocket.prototype.onerror = function(event) {};
    /** An event listener to be called when total closed */
    ReconnectingWebSocket.prototype.ontotalclose = function(event) {};

    ReconnectingWebSocket.prototype.socketHeartbeat = null;

    /**
     * Whether all instances of ReconnectingWebSocket should log debug messages.
     * Setting this to true is the equivalent of setting all instances of ReconnectingWebSocket.debug to true.
     */
    ReconnectingWebSocket.debugAll = false;

    ReconnectingWebSocket.CONNECTING = WebSocket.CONNECTING;
    ReconnectingWebSocket.OPEN = WebSocket.OPEN;
    ReconnectingWebSocket.CLOSING = WebSocket.CLOSING;
    ReconnectingWebSocket.CLOSED = WebSocket.CLOSED;

    return ReconnectingWebSocket;
});


utools-chatroom/static/js/index.js
$(document).ready(function($) {

Date.prototype.Format = function(fmt)
{ //author: meizz
	var o = {
	"M+" : this.getMonth()+1,                 //月份
	"d+" : this.getDate(),                    //日
	"h+" : this.getHours(),                   //小时
	"m+" : this.getMinutes(),                 //分
	"s+" : this.getSeconds(),                 //秒
	"q+" : Math.floor((this.getMonth()+3)/3), //季度
	"S"  : this.getMilliseconds()             //毫秒
	};
	if(/(y+)/.test(fmt))
		fmt=fmt.replace(RegExp.$1, (this.getFullYear()+"").substr(4 - RegExp.$1.length));
	for(var k in o)
		if(new RegExp("("+ k +")").test(fmt))
			fmt = fmt.replace(RegExp.$1, (RegExp.$1.length==1) ? (o[k]) : (("00"+ o[k]).substr((""+ o[k]).length)));
	return fmt;
}

if (!window.MozWebSocket && !window.WebSocket ){
	alert("你的浏览器不支持websocket.");
	return
}

var needNickname = function(){
	console.log("needNickname");
	$("#nickname-error").hide();
	$("#login-modal").modal({
		keyboard:false,
		backdrop:"static"
	});
	$('#login-modal').on('shown.bs.modal', function () {
		$("#nickname-edit").focus();
	})
};

var setNicknameError = function(_error_message){
	console.log("setNicknameError : " + _error_message);
	$("#nickname-error-message").text(_error_message);
	$("#nickname-error").show();
	$('#nickname-edit').focus();
};

var setNicknameSuccess = function(_new_nickname){
	console.log("setNicknameSuccess : " + _new_nickname);
	$("#login-modal").modal("hide");
	$("#my-nickname").text(_new_nickname);
	$("input[id=input-edit]").focus();
};

var sayMessageError = function(_error_message){
	console.log("sayMessageError: "+sayMessageError);
};

var serverMessage = function(_message){
	console.log("serverMessage: "+_message);
	chat_Socket.serverMessage(_message, chat_Socket.getLocalHMS());
};

var messageError = function (_message) {
	console.log("messageError: "+_message);
	chat_Socket.messageError(_message, chat_Socket.getLocalHMS());
}

var userMessage = function(_message){
	_nick_name = _message['username'];
	_content = _message['content'];
	// _nick_name = _message.split("$")[0];
	// _content = _message.split("$")[1];
	console.log(_nick_name + " say: " + _content);
	chat_Socket.userMessage(_nick_name, _content, chat_Socket.getLocalHMS());
};

var userImage = function (_message) {
	_nick_name = _message['username'];
	_content = _message['content'];
	console.log(_nick_name + " send: " + _content);
	chat_Socket.userImage(_nick_name, _content, chat_Socket.getLocalHMS());
}

var userList = function(_list){
	console.log("userList: "+_list);
	// _list = _list.split("$");
	chat_Socket.userList(_list);
};

var userJoin = function(_nick_name){
	console.log("userJoin: "+_nick_name);
	if( _nick_name == $("#my-nickname").text())
		return
    // chat_Socket.addUserToList(_nick_name);
    // chat_Socket.updateListCount();
    chat_Socket.serverMessage(_nick_name+" 加入了聊天室 (*^__^*) …… ",chat_Socket.getLocalHMS());
};

var userQuit = function(_nick_name){
	console.log("userQuit: "+_nick_name);
    // chat_Socket.removeUserFromList(_nick_name);
    // chat_Socket.updateListCount();
    chat_Socket.serverMessage(_nick_name+" 离开了聊天室 ( ⊙ o ⊙ ) … ",chat_Socket.getLocalHMS());
};


var websocket_url = (location.protocol === "https:" ? 'wss://' : 'ws://') +
	location.hostname + (location.port ? ':'+location.port: '') + "/websocket" +
	location.pathname;

socket = new ReconnectingWebSocket("ws://39.105.91.194:8090/websocket/", null, {"maxReconnectAttempts": 1});


ReconnectingWebSocket.prototype.onmessage = function(event) {
	data = JSON.parse(event.data);
	action = data['action'];
	message = data['message'];
	switch (action) {
		case "needNickname": needNickname();
							break;
		case "setNicknameError": setNicknameError(message);
							break;
		case "setNicknameSuccess": setNicknameSuccess(message);
							break;
		case "sayMessageError": sayMessageError(message);
							break;
		case "serverMessage": serverMessage(message);
							break;
		case "userMessage": userMessage(message);
							break;
		case "messageError": messageError(message);
							break;
		case "userImage": userImage(message);
							break;
		case "userList": userList(message);
							break;
		case "userJoin": userJoin(message);
							break;
		case "userQuit": userQuit(message);
							break;
		case "heartbeat": break
		default: console.log("Error:"+action);
	}
};


ReconnectingWebSocket.prototype.onopen = function(event) {
	chat_Socket.hideReconnection()
	chat_Socket.serverMessage("你已进入聊天室", chat_Socket.getLocalHMS());
	ReconnectingWebSocket.prototype.socketHeartbeat = setInterval(function()
	{
		socket.send(JSON.stringify({"action": "heartbeat", "message": "ping"}));
	}, 10000); //10000 milliseconds = 10 seconds
};

ReconnectingWebSocket.prototype.onclose = function(event) {
	clearInterval(ReconnectingWebSocket.prototype.socketHeartbeat);
	chat_Socket.serverMessage("聊天室已关闭。。。", chat_Socket.getLocalHMS());
};

ReconnectingWebSocket.prototype.ontotalclose = function(event) {
	chat_Socket.showReconnection();
	$("#login-modal").modal("hide");
};

chat_Socket = {
	init: function(){
		this.initEmotion();
		this.initMessage();
		this.sendNickname();
	},
	initEmotion: function(){
		QxEmotion($('#emotion-btn'), $('#input-edit'));
	},
	initMessage: function(){
		var self = this;
		$("#send-message").on("click", function(event){
			event.preventDefault();
			self.applyMessage();
		});
		$("#msg-list-body").on("click", "#reconnection", function(event){
			event.preventDefault();
			socket.open();
		});
		$("form").submit(function(event) {
			event.preventDefault();
			self.applyMessage();
		});
	},
	applyMessage: function(){
		var content = $("input[id=input-edit]").val();
		if(content){
			this.sendMessage(content)
			$("input[id=input-edit]").val('');
		}
	},
	sendMessage: function(content){
		socket.send(JSON.stringify({"action":"message", "message":content}));
	},
	sendImage: function(content){
		socket.send(JSON.stringify({"action":"image", "message":content}));
	},
	sendNickname: function(){
		var self = this;
		$("#nickname-edit").keydown(function(event){
			if( event.keyCode == 13)
				self.applyNickname();
		});
		$("#applyNicknameBtn").on('click', function(){
			self.applyNickname();
		})
	},
	applyNickname: function(){
		var nickname_edit = $("#nickname-edit");
		var nickname_error = $("#nickname-error");
		var nickname = nickname_edit.val();
		if (nickname.trim() == ""){
			$("#nickname-error-message").text("请填写昵称。");
			nickname_error.show();
			nickname_edit.focus();
			return;
		}
		if (nickname.length > 20 ){
			$("#nickname-error-message").text("昵称过长，长度应小于20个字符。");
			nickname_error.show();
			nickname_edit.focus();
			return;
		}
		socket.send(JSON.stringify({"action":"setNickname", "message":nickname}));
	},
	chatBodyToBottom: function(){
		// var chat_body = $('.main #chat-board');
		var chat_body = document.getElementById('chat-board');
		chat_body.scrollTop = chat_body.scrollHeight + 10000;
	},
	serverMessage: function(_content, _time){
		_content = QxEmotion.Parse(_content);
		var messages = $(".msg-list-body");
		var message = '<div class="text-center sys-message">\
							<span class="sys-tip"> 系统消息: '+_content+'    &nbsp;'+_time+'</span>\
						</div>';
		messages.append(message);
		this.chatBodyToBottom();
	},
	hideReconnection: function(){
		$("button.reconnection").hide();
	},
	showReconnection: function(){
		var messages = $(".msg-list-body");
		var message = '<button type="button" id="reconnection" class="center-block btn btn-success reconnection">点击重连</button>'
		messages.append(message);
		this.chatBodyToBottom();
	},
	messageError: function(_content, _time){
		_content = QxEmotion.Parse(_content);
		var messages = $(".msg-list-body");
		var message = '<div class="text-center sys-message">\
							<span class="error-msg"> 发送异常: '+_content+'    &nbsp;'+_time+'</span>\
						</div>';
		messages.append(message);
		this.chatBodyToBottom();
	},
	userMessage: function(_nick_name, _content, _time){
		_content = QxEmotion.Parse(_content);
		var messages = $(".msg-list-body");
		var nickname = $("#my-nickname").text();
		if (_nick_name != nickname) {
			utools.showNotification(_nick_name+':'+_content)
		}
		var textalign = _nick_name == nickname ? "text-right" : "";
		var label =  _nick_name == nickname ? "label-success" : "label-info";
		var message = '<div class="message jumbotron '+textalign+' ">\
							<div class="message-head">\
								<span class="label '+label+'"><span class="glyphicon glyphicon-user"></span>&nbsp;&nbsp;'+_nick_name+'</span>&nbsp;&nbsp;\
								<span class="label label-default"><span class="glyphicon glyphicon-time"></span>&nbsp;&nbsp;'+_time+'</span>&nbsp;&nbsp;\
							</div>\
							<div class="message-body">\
								'+_content+'\
							</div>\
						</div>';
		messages.append(message);
		this.chatBodyToBottom();
	},
	userImage: function(_nick_name, _content, _time){
		_content = QxEmotion.Parse(_content);
		var messages = $(".msg-list-body");
		var nickname = $("#my-nickname").text();
		var textalign = _nick_name === nickname ? "text-right" : "";
		var label =  _nick_name === nickname ? "label-success" : "label-info";
		var message = '<div class="message jumbotron '+textalign+' ">\
							<div class="message-head">\
								<span class="label '+label+'"><span class="glyphicon glyphicon-user"></span>&nbsp;&nbsp;'+_nick_name+'</span>&nbsp;&nbsp;\
								<span class="label label-default"><span class="glyphicon glyphicon-time"></span>&nbsp;&nbsp;'+_time+'</span>&nbsp;&nbsp;\
							</div>\
							<div class="message-body">\
								<img onload=loaded_image() class="message-image" src="'+_content+'" alt="image" > \
							</div>\
						</div>';
		messages.append(message);
		this.chatBodyToBottom();
	},
	userList: function(_list){
		$(".list .list-group").html("");
		if (_list[0] != "")
			for(var i = 0; i < _list.length; i++ ) {
				this.addUserToList(_list[i]);
			};
		this.updateListCount();
	},
	addUserToList: function(_nick_name){
		$(".list .list-group").append('<li class="list-group-item">'+_nick_name+'</li>');
	},
	removeUserFromList: function(_nick_name){
		$(".list .list-group li").each(function(){
			if(_nick_name == $(this).text()){
				$(this).remove();
			}
		});
	},
	updateListCount: function(){
		var list_count = $(".list .list-group").find("li").length;
		$("#list-count").text("当前在线人数："+list_count+"人");
	},
	getLocalHMS: function(){
		return new Date().Format("yyyy-MM-dd hh:mm:ss");
	},
}

chat_Socket.init();


    $("div#mainArea").dropzone({
        url: "http://39.105.91.194:8090/upload_image",
        paramName: "image",
        clickable: false,
		// 限制了也没有友好的提示
		// acceptedFiles: "image/*",
        addedfile: function (file) {
            return false;
        },
        success: function (file, response) {
            if (response.code === 0){
                chat_Socket.sendImage(response.data.url)
            } else {
                console.log("response: " + response.message)
                chat_Socket.messageError("文件上传失败", chat_Socket.getLocalHMS());
            }
        },
        dragenter: function (e) {
            $("#cover").removeClass("cover-empty");
            this.element.classList.add("dz-drag-hover");
        },

        dragover: function (e) {
            $("#cover").removeClass("cover-empty");
            this.element.classList.add("dz-drag-hover");
        },
        dragleave: function (e) {
            this.element.classList.remove("dz-drag-hover");
        },
        drop: function (e) {
            $("#cover").addClass("cover-empty");
            this.element.classList.remove("dz-drag-hover");
        },

    });

    document.getElementById("cover").ondragover = function () {
        $("#cover").addClass("cover-empty");
        return false;
    }
    document.getElementById("cover").ondragleave = function () {
        $("#cover").addClass("cover-empty");
        return false;
    }
    // 全局阻止放下图片事件，禁止浏览器默认行为
    window.ondrag=function(e) {
        e.preventDefault();
        return false
    }
    window.ondragover=function(e) {
        e.preventDefault();
        return false
    }
    window.ondrop=function(e) {
        e.preventDefault();
        return false
    }

    function send_file(file) {
        var fd = new FormData();
        fd.append("image", file)

        $.ajax({
            url: 'http://39.105.91.194:8090/upload_image',
            type: 'post',
            data: fd,
            contentType: false,
            processData: false,
            success: function(response){
                if(response.code === 0){
                    chat_Socket.sendImage(response.data.url)
                }else{
                    console.log("response" + response.message)
                    chat_Socket.messageError("文件上传失败", chat_Socket.getLocalHMS());
                }
            },
        });
    }

    document.addEventListener('paste', function (event) {
        var items = event.clipboardData && event.clipboardData.items;
        var file = null;
        if (items && items.length) {
            // 检索剪切板items
            for (var i = 0; i < items.length; i++) {
                if (items[i].type.indexOf('image') !== -1) {
                    file = items[i].getAsFile();
                    break;
                }
            }
        }

        if (file === null){
            return false;
        }
        var extIndex = file.name.lastIndexOf(".");
        var extValue = file.name.substring(extIndex+1).toLowerCase();
        if (extValue !== "jpg" && extValue !== "jpeg" && extValue !== "png" && extValue !== "gif"){
            chat_Socket.serverMessage("非图片文件", chat_Socket.getLocalHMS());
            return false;
        }
        send_file(file)

    });

});

function loaded_image() {
	chat_Socket.chatBodyToBottom();
}
